doc/SAI-Proposal-PoE.md

```diff
@@ -0,0 +1,591 @@
+2. The PoE software provides an interface to configure (program) and query the PoE controller and its internal peripherals such as ports etc
+3. Users can see PoE information and configure PoE features according to their needs
+
+# PoE Hardware Design Example
+
+![poe-physical-design](figures/PoE_Physical_Design.png)
+
+# SAI POE Components
+
+## New PoE SAI objects
+
+Three new SAI objects are introduced.
+
+### PoE Device ID
+
+This object is used to hold the global configuration and PoE hardware status. In many cases, one PoE device instance will have access to all the PoE hardware and components.
+
+### PoE PSE ID
+
+This object is used to access data from the Power Sourcing Equipment devices in the board. Such as PSE temperatures, versions, and statuses. There can be multiple PSEs in one board.
+
+### PoE Port ID
+
+This object is used to access data on a port level. PoE ports supply power using an ethernet cable. PoE port object gives access to read data and configure these ports.
+
+## New SAI Switch Type PoE
+
+A new abstraction type for PoE is created which allows multiple vendors to implement their own PoE software and hardware solutions.
+
+```
+/**
+ * @brief Attribute data for #SAI_SWITCH_ATTR_TYPE
+ */
+typedef enum _sai_switch_type_t
+{
+    /** Switch type is Switching Network processing unit */
+    SAI_SWITCH_TYPE_NPU,
+
+    /** Switch type is PHY */
+    SAI_SWITCH_TYPE_PHY,
+
+    /** Switch type is VOQ based NPU */
+    SAI_SWITCH_TYPE_VOQ,
+
+    /** Switch type is Fabric switch device */
+    SAI_SWITCH_TYPE_FABRIC,
+
+    /** Switch type is PoE (Power over Ethernet) */
+    SAI_SWITCH_TYPE_POE,
+
+} sai_switch_type_t;
+```
+
+## Example of how to initialize  PoE
+
+An example of how to initlizate PoE and access its components.
+
+### Create a PoE Device Object
+
+The PoE device object is used to access and configure board-wide system information and operations. Information about the kind of PoE hardware should be provided on the NOS level and used to initalize the driver.
+
+```
+    /**
+     * @brief Device Information for switch initialization.
+     *
+     * Hardware information format is based on POE SAI implementations by vendors.
+     * Format is vendor specific.
+     * Example: Like PCI location, I2C address, UART etc.
+     * In case of NULL, use the default PoE associated switch id.
+     *
+     * @type char
+     * @flags CREATE_ONLY
+     * @default ""
+     */
+    SAI_POE_DEVICE_ATTR_HARDWARE_INFO = SAI_POE_DEVICE_ATTR_START,
+```
+
+```
+/** 
+ * Create a PoE device object
+ */
+  
+/* Allows the PoE software to initialize itself based on the PoE hardware */
+attr_list[0].id = SAI_POE_DEVICE_ATTR_HARDWARE_INFO;
+strncpy(attr.value.chardata, "integrated-mcu", sizeof("integrated-mcu");  /* string is only relevant to the driver, can be anything*/ 
+
+attr_count = 1;
+sai_create_poe_device_fn(&poe_device_id, switch_id, attr_count, attr_list);
```

Comment:
A Switch(MAC) device can have multiple ports and these ports can span across multiple PoE controllers like as shown below:
              ports 1 ... K   |---------- PoE Controller1 
MAC --- ports K ... M  |---------- PoE Controller2
              ports  M ... N |---------- PoE Controller3
                                and so on
Like some 'J' PoE controllers are serving 1 ... N ports of MAC(Switch) device.
We need the switch Id and the PoE controller (instance) Id here for initializing all the PoE devices under a switch.

Reply:
In our example, we have a microcontroller chip that runs PoE firmware (manages all PSE controllers), and the system only communicates with the MCU and not directly with all the PoE components.
If you wish to control each PoE controller individually, please create a PoE device for each relevant one.

Reply:
Ok.